app/controllers/ally_controller.rb

```diff
@@ -0,0 +1,177 @@
+#
+# Copyright (C) 2011 Instructure, Inc.
+#
+# This file is part of Canvas.
+#
+# Canvas is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, version 3 of the License.
+#
+# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Affero General Public License along
+# with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+# @API Ally
+# API for access Ally data
+class AllyController < ApplicationController
+
+  # @API Get Ally information
+  # Returns the Ally integration information
+  #
+  # @example_request
+  #
+  #   curl 'https://<canvas>/api/v1/ally' \
+  #         -H 'Authorization: Bearer <token>'
+  #
+  # @example_response
+  #
+  #  { "enabled": true, "clientId": 47831, "baseUrl": "https://prod.ally.ac" }
+  #
+  def enabled
+    get_context
+    ally_settings = get_ally_settings
+    data = {
+      :enabled => ally_settings[:enabled],
+      :clientId => ally_settings[:client_id],
+      :baseUrl => ally_settings[:base_url]
+    }
+    render :json => data
+  end
+
+  # @API Create a signed request to the Ally API
+  # Get the OAuth signature so a request to the Ally API can be made
+  #
+  # @example_request
+  #
+  #   curl 'https://<canvas>/api/v1/courses/<course_id>/ally/sign?http_method=<GET|POST>&http_path=<path>&http_parameters=<url encoded parameters>' \
+  #         -H 'Authorization: Bearer <token>'
+  #
+  # @example_response
+  #
+  #  { "clientId": 47831, "baseUrl": "https://prod.ally.ac", "path": "<path>", "header": "87f82d487f78ffae4dd8b8edaa2547d7", "body": "<urlencoded body>" }
+  #
+  def sign
+    get_context
+    request = get_signed_request_object()
+    if (request)
+      ally_settings = get_ally_settings()
+      data = {
+        :clientId => ally_settings[:client_id],
+        :baseUrl => ally_settings[:base_url],
+        :path => request.path,
+        :header => request.to_hash['authorization'][0],
+        :body => request.body,
+      }
+      render :json => data
+    end
+  end
+
+  ##API Proxy a request to the Ally API
+  # Proxy data from the Ally REST API
+  #
+  # @example_request
+  #
+  #   curl 'https://<canvas>/api/v1/courses/<course_id>/ally/proxy?http_method=<GET|POST>&http_path=<path>&http_parameters=<url encoded parameters>' \
+  #         -H 'Authorization: Bearer <token>'
+  def proxy
+    get_context
+    request = get_signed_request_object()
+    if request
+      ally_settings = get_ally_settings()
+      uri = URI.parse(ally_settings[:base_url])
+      http = Net::HTTP.new(uri.host, uri.port)
```

Comment:
This looks like it's ignoring the scheme? Does it come out in the scheme by the uri parser? If not, what determines if it's secure?